### PR TITLE
Escaping and better .htaccess hdd sample code

### LINKED ISSUE #157: hdd.htaccess
cachify/inc/setup/cachify.hdd.htaccess.php:

1.) main rules: for every non cachify file (e.g js, css, jpg...) existing of this files inside cachify path is checked!

2.) I miss some escapings in rewrite conditions

My proposal for the # Main Rules:

```
# Main Rules
RewriteCond %{ENV:CACHIFY_DIR} !^$
RewriteCond %{REQUEST_METHOD} !=POST
RewriteCond %{QUERY_STRING} ^$
RewriteCond %{REQUEST_URI} !^\/(wp\-admin|wp\-content\/cache)\/.*
RewriteCond %{HTTP_COOKIE} !(wp\-postpass|wordpress_logged_in|comment_author)_
RewriteCond %{DOCUMENT_ROOT}/wp-content/cache/cachify/%{ENV:CACHIFY_HOST}%{ENV:CACHIFY_DIR}index.html%{ENV:CACHIFY_SUFFIX}-f
RewriteRule ^(.*) /wp-content/cache/cachify/%{ENV:CACHIFY_HOST}%{ENV:CACHIFY_DIR}index.html%{ENV:CACHIFY_SUFFIX} [L]
```

### LINKED ISSUE #200: POST-Request to / receive a cached response (HDD-Cache)
With the Apache-rewrite rules provided by the plugins Setup tab POST request to / receive a (hdd) cached answer.

I'm not entirely sure what exactly goes wrong but it seems that after the final WordPress rule RewriteRule . /index.php [L] is applied, all the rewrite rules are checked again (which apparently is how apache works. did not know that before). inside the subrequest the REQUEST_METHOD seems not to match the initial one which leads to a cached response being delivered for a POST request to /

```
1 [Wed Oct 07 13:35:09.120411 2020] [rewrite:trace3] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca020e0a0/initial] [perdir /var/www/html/] applying pattern '.' to uri '', referer: https://localhost/
2 [Wed Oct 07 13:35:09.120417 2020] [rewrite:trace1] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca020e0a0/initial] [perdir /var/www/html/] pass through /var/www/html/, referer: https://localhost/
3 [Wed Oct 07 13:35:09.120473 2020] [rewrite:trace3] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] strip per-dir prefix: /var/www/html/index.php -> index.php, referer: https://localhost/
4 [Wed Oct 07 13:35:09.120490 2020] [rewrite:trace3] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] applying pattern '.*' to uri 'index.php', referer: https://localhost/
...
5 [Wed Oct 07 13:35:09.120627 2020] [rewrite:trace4] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] RewriteCond: input='GET' pattern='!=POST' => matched, referer: https://localhost/
6 [Wed Oct 07 13:35:09.120634 2020] [rewrite:trace4] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] RewriteCond: input='' pattern='=""' => matched, referer: https://localhost/
7 [Wed Oct 07 13:35:09.120640 2020] [rewrite:trace4] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] RewriteCond: input='/index.php' pattern='!^/(wp-admin|wp-content/cache)/.*' => matched, referer: https://localhost/
...
8 [Wed Oct 07 13:35:09.120683 2020] [rewrite:trace2] [pid 61:tid 140516971439872] mod_rewrite.c(483): [client 192.168.96.1:55700] 192.168.96.1 - - [localhost/sid#7fccb5d18420][rid#7fcca02020a0/subreq] [perdir /var/www/html/] rewrite 'index.php' -> '/wp-content/cache/cachify/https-localhost/index.html.gz', referer: https://localhost/
```

line 1: is the final WordPress rule RewriteRule . /index.php [L] 
which results in a subrequest (lines 2 and 3)
line 4 is the first cachify rule RewriteRule .* - [E=CACHIFY_HOST:https-%{HTTP_HOST}] this time applied to the subrequest to index.php
line 5 shows that for the subrequest the method is GET
6 & 7 are the cachify conditions
RewriteCond %{REQUEST_URI} !^/(wp-admin|wp-content/cache)/.*
RewriteCond %{HTTP_COOKIE} !(wp-postpass|wordpress_logged_in|comment_author)_
and in line 8 it gets rewritten to the cached path

I "fixed" it by prepending RewriteCond %{REQUEST_URI} /$ to the Main Rules block which makes the whole "set subdirectory" block useless. I'm not sure under which circumstances the condition RewriteCond %{REQUEST_URI} ^$ is true so im not sure if my "fix" breaks other setups.

### PR DESCRIPTION
**Pull request by timse201**
closes #157

### CODE REVIEW

**Comment by florianbrinkmann** [comment]
Hi @timse201, I tried to verify that the changes fix the issue with the cached post request, but I still get the cached version for post requests with the changed `.htaccess`. Could you tell me the steps how I can test that? Thanks!

**Comment by florianbrinkmann** [comment]
The PR does not fix #200, I will open a PR for that later today. I am not sure about the other changes, is escaping needed here, does that have any advantages over leaving the `.htaccess` example as it is?

File: inc/setup/cachify.hdd.htaccess.php [hunk 1]
```diff
@@ -19,18 +19,14 @@
...
 {{GZIP}}
   # Main Rules
   RewriteCond %{HTTP_ACCEPT} .*text/html.*
+  RewriteCond %{REQUEST_URI} /$
   RewriteCond %{REQUEST_METHOD} GET
-  RewriteCond %{QUERY_STRING} =""
-  RewriteCond %{REQUEST_URI} !^/(wp-admin|wp-content/cache)/.*
-  RewriteCond %{HTTP_COOKIE} !(wp-postpass|wordpress_logged_in|comment_author)_
+  RewriteCond %{QUERY_STRING} ^$
+  RewriteCond %{REQUEST_URI} !^\/(wp\-admin|wp\-content\/cache)\/.*
```
Comment (stklcode) [inline inc/setup/cachify.hdd.htaccess.php:28]: Is there any reason to escape the characters here?

`/` is not used as a delimiter in this context and `-` is no reserved character outside character groups 🤔

File: inc/setup/cachify.hdd.htaccess.php [hunk 2]
```diff
@@ -19,18 +19,14 @@
...
-  RewriteRule .* - [E=CACHIFY_DIR:/]
 {{GZIP}}
   # Main Rules
   RewriteCond %{HTTP_ACCEPT} .*text/html.*
+  RewriteCond %{REQUEST_URI} /$
   RewriteCond %{REQUEST_METHOD} GET
-  RewriteCond %{QUERY_STRING} =""
-  RewriteCond %{REQUEST_URI} !^/(wp-admin|wp-content/cache)/.*
-  RewriteCond %{HTTP_COOKIE} !(wp-postpass|wordpress_logged_in|comment_author)_
+  RewriteCond %{QUERY_STRING} ^$
```
Comment (stklcode) [inline inc/setup/cachify.hdd.htaccess.php:27]: I don’t have strong feelings about the notation, but is there any benefit in using an empty RegEx over comparison against an empty string?

**Comment by stklcode** [comment]
Closing this PR in favor of #302 and other refactoring around that area. Real issues should be all resolved by now. Feel free to reopen this in case we missed something.

**Pull request closed without merging**